ksqldb-parser/src/main/java/io/confluent/ksql/parser/SyntaxErrorValidator.java

```diff
@@ -127,16 +135,77 @@ public void syntaxError(
       return;
     }
 
-    if (offendingSymbol instanceof Token && isKeywordError(
-        message, ((Token) offendingSymbol).getText())) {
-      //Checks if the error is a reserved keyword error
-      final String tokenName = ((Token) offendingSymbol).getText();
-      final String newMessage =
-          "\"" + tokenName + "\" is a reserved keyword and it can't be used as an identifier."
-              + " You can use it as an identifier by escaping it as \'" + tokenName + "\' ";
-      throw new ParsingException(newMessage, e, line, charPositionInLine);
+    if (offendingSymbol instanceof Token) {
+      final StringBuilder sb = new StringBuilder();
+      sb.append("Syntax Error\n");
+      if (isKeywordError(message, ((Token) offendingSymbol).getText())) {
+        //Checks if the error is a reserved keyword error
+        final String tokenName = ((Token) offendingSymbol).getText();
+        sb.append(
+            String.format("\"%s\" is a reserved keyword and it can't be used as an identifier."
+                + " You can use it as an identifier by escaping it as '%s' ",
+                tokenName, tokenName));
+      } else if (message.contains("expecting")) {
+        sb.append(
+            getExpectations(message, ((Token) offendingSymbol).getText())
+        );
+      } else if (e instanceof NoViableAltException) {
+        sb.append(
+            String.format(
+                "Syntax error at or near '%s' at line %d:%d",
+                ((Token) offendingSymbol).getText(), line, charPositionInLine + 1
+            )
+        );
+      } else {
+        sb.append(message);
+      }
+      throw new ParsingException(sb.toString(), e, line, charPositionInLine);
     } else {
       throw new ParsingException(message, e, line, charPositionInLine);
     }
   }
+
+  private String getExpectations(
+      final String message,
+      final String offendingToken) {
+    final StringBuilder output = new StringBuilder();
+    final String expectingStr = message.split("expecting ")[1];
+    // If the command is mistyped, find the most similar command and show it
+    if (isCommand(expectingStr)) { // this is a command typo
+      output.append(String.format("Unknown statement '%s'%n", offendingToken));
+      output.append(
+              String.format("Did you mean '%s'?", getMostSimilar(offendingToken)));
+    // In case of missing closing brackets or parentheses
+    } else if (message.contains("EOF")) {
+      output.append("Syntax error at or near \";\"\n");
+      output.append("'}', ']', or ')' is missing");
+    // If the expecting list generated by Antlr is a short one, show it to user
+    } else if (expectingStr.split(",").length <= 3) {
+      output.append(String.format("Expecting %s", expectingStr));
```

Comment:
This could be in another PR, but the expecting list generated by Antlr is often pretty bad. For example,

```
ksql> describe functiii concat;
line 1:19: Syntax Error
Expecting ';'
Statement: describe functiii concat;
Caused by: line 1:19: Syntax Error
Expecting ';'

ksql> create stream foo a select * from bar;
line 1:19: Syntax Error
Expecting {';', '(', 'WITH'}
Statement: create stream foo a select * from bar;
Caused by: line 1:19: Syntax Error
Expecting {';', '(', 'WITH'}
Caused by: org.antlr.v4.runtime.InputMismatchException

```